### ISSUE TITLE
Detaching roles (and permissions) should flush the cache.

### ISSUE THREAD

**Issue by niknetniko** [romanbican/roles#90]
When using `detachAllRoles()` and similar functions, the `$roles` property should be emptied (or at least provide an option to do so). Otherwise, the following is not possible:
```php
$isAdmin = $user->is('admin');

$user->detachAllRoles();

//Role ID's from html form with checkboxes
$roles = $request->get('role');

foreach ($roles as $role) {
    $roleObject = Role::find($role);
    $user->attachRole((int) $role);
}
```

I do this because I have a form with checkboxes where the admins can select the roles an user has. However, this currently doesn't work, because existing roles are not re-added. The workflow is this:

1. The first line checks if the user is an admin. This populates the `$roles` property in the `HasRoleAndPermission` trait.
2. The `detachAllRoles()` removes all roles from the user, but does not clear the `$roles` property.
3. When re-adding a role, the `attachRole($id)` function checks if the role exists or not before adding it. However, this uses the `$roles` property, which still contains the old roles. In reality all roles have been removed. Since `$roles` already has the role, no new roles are attached, resulting in an user without roles.

If you want, I can make a pull request with the modified functions. `detachAllRoles()` for example would be:
```php
/**
 * Detach all roles from a user.
 *
 * @return int
 */
public function detachAllRoles()
{
    $this->roles = null;
    return $this->roles()->detach();
}
```

**Comment by romanbican** [comment]
Yes, please make a pull request.

**Comment by niknetniko** [comment]
OK, I will do it first thing tomorrow.

**Comment by niknetniko** [comment]
Closed with #93 :+1:
